Add complete todo app with Embed Workflow integration demo

AntonyWilliam opened the pull request:
This commit introduces a comprehensive beginner-level demo showcasing how to integrate Embed Workflow into a Next.js todo
  application with email and Slack notifications.

  Key Features:
  - Complete Next.js todo app with task creation and completion
  - Embedded workflow builder UI integration
  - Dual notification system (email + Slack)
  - JWT authentication for secure workflow access
  - Environment variable configuration
  - Fallback notification system for reliability

  Documentation Includes:
  - Step-by-step setup instructions with screenshots
  - API key configuration guide
  - Email and Slack workflow configuration
  - JSON block examples for rich Slack messaging
  - Variable usage examples
  - Troubleshooting and security notes

  Technical Implementation:
  - Next.js 12 with React 17 and Tailwind CSS
  - Embed Workflow SDK integration (v1.43.0)
  - Environment-based API key management
  - Local storage for task persistence
  - Comprehensive error handling

  Demo Organization:
  - Media folder with demo GIF and setup screenshots
  - Proper gitignore for development artifacts
  - Clean project structure for easy understanding
  - Ready-to-run example with clear setup steps

  This demo serves as a complete reference for developers wanting to integrate workflow automation into their applications.

AntonyWilliam committed:
Add complete todo app with Embed Workflow integration demo

  This commit introduces a comprehensive beginner-level demo showcasing how to integrate Embed Workflow into a Next.js todo
  application with email and Slack notifications.

  Key Features:
  - Complete Next.js todo app with task creation and completion
  - Embedded workflow builder UI integration
  - Dual notification system (email + Slack)
  - JWT authentication for secure workflow access
  - Environment variable configuration
  - Fallback notification system for reliability

  Documentation Includes:
  - Step-by-step setup instructions with screenshots
  - API key configuration guide
  - Email and Slack workflow configuration
  - JSON block examples for rich Slack messaging
  - Variable usage examples
  - Troubleshooting and security notes

  Technical Implementation:
  - Next.js 12 with React 17 and Tailwind CSS
  - Embed Workflow SDK integration (v1.43.0)
  - Environment-based API key management
  - Local storage for task persistence
  - Comprehensive error handling

  Demo Organization:
  - Media folder with demo GIF and setup screenshots
  - Proper gitignore for development artifacts
  - Clean project structure for easy understanding
  - Ready-to-run example with clear setup steps

  This demo serves as a complete reference for developers wanting to integrate workflow automation into their applications.

amrani commented:
Taking a step back.. we should always get potential users to the Aha moment as fast as possible. 
  - Our  Aha moment is when the user sees the embedded application in their local dev environment. 

Consider restructuring the demo to the following:

1. Navigate to Project Directory
2. Install Dependencies
3. Install Embed Workflow / Embed the workflow builder / etc (previously named "Set Environment Variables")
  - Dependency: You have an Embed Workflow account. Get the API Key's
4. Run the Development Server
  - This is the Aha moment. As long as they get here... we win!
5. Configure Embed Workflow


Regarding the Configure Embed Workflow section:

for configuration, let's use `YAML` code blocks so they can clearly see the values. Plus, we are actively working on a copy paste feature so they can take your entire code block and generate the settings.

1. Create Task Created Trigger

```YAML
  event: "todo_list_item_created"
  icon:
    type: "bolt"
    background_color: "slate"
  groups: []
  data_input_schema:
    - type: "String"
      required: true
      variable: "task_id"
      data_path: "task_id"
      display_label: "Task ID"
    - type: "String"
      required: true
      variable: "task_name"
      data_path: "task_name"
      display_label: "Task Name"
    - type: "String"
      required: true
      variable: "task_description"
      data_path: "task_description"
      display_label: "Task Description"
    - type: "Date"
      required: true
      variable: "created_at"
      data_path: "created_at"
      display_label: "Created At"
    - type: "String"
      required: true
      variable: "email_list"
      data_path: "email_list"
      display_label: "Email List"
  title: "Task Created"
  description: "A task was created"
```

2. Display the `YAML` for this one too. (*pro tip; they can duplicate the trigger and edit*)

3. Create workflows
  - At this point, they already embedded the user. We can have them to build the workflow in their App!

AntonyWilliam committed:
Update

AntonyWilliam committed:
Summary of Changes Made

Reordered Quick Start steps to prioritize seeing the embedded workflow builder:

  1. Navigate to Project Directory
  2. Install Dependencies
  3. Set Up Embed Workflow (moved environment setup earlier)
  4. Run Development Server
  5. Experience Your Embedded Automation Platform (the Aha moment)
  6. You're Ready to Build!

Key Improvements

✅ Moved API key setup before configuration - users get their keys and run the app first

✅ Added step 6 celebration - "You're Ready to Build!" as finale

✅ Converted triggers to YAML
  format - copy-paste ready configurations

✅ Updated workflow creation - emphasizes using the embedded builder they just saw

YAML Implementation

✅ Task Created Trigger - Full YAML configuration block

✅ Task Completed Trigger - YAML with duplication tip

✅ Copy-paste ready - Prepared for upcoming copy-paste feature

  Flow Achievement

✅ Fast path to embedded builder - Users see the embedded workflow UI in ~5 minutes

✅ Configuration comes after - Triggers and workflows setup happens after the "win"

✅ Embedded  workflow emphasis - Users build workflows in their own app, not external dashboard

AntonyWilliam committed:
Update

AntonyWilliam committed:
Update

AntonyWilliam committed:
update

amrani commented on beginner/todo-slack-notifications/pages/workflows/index.js:
```diff
@@ -0,0 +1,184 @@
+              media="screen"
+              href={`https://cdn.ewf.to/ewf-${ewfVersion}.css`}
+            />
+            {loadError ? (
```
you can remove the entire section. You have the back to tasks button in the nav and the embedded component actually has a similar retry UI

amrani commented on beginner/todo-slack-notifications/pages/index.js:
```diff
@@ -0,0 +1,563 @@
+            Tasks
+          </button>
+          <button
+            onClick={() => setActiveTab('workflows')}
```
instead of loading tab content, you can also navigate to the workflows page. Less might be better here

amrani commented on beginner/todo-slack-notifications/.env.template:
```diff
@@ -0,0 +1,16 @@
+
+# UI Version - Check latest version at: https://embedworkflow.com/ui-version
+EMBED_WORKFLOW_UI_VERSION=latest
+NEXT_PUBLIC_EMBED_WORKFLOW_UI_VERSION=latest
```
i recommend using "x.x.x" instead of "latest". Someone might think "latest" will auto pull the latest version and not update this env

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,667 @@
+EMBED_WORKFLOW_SK=your_secret_key_here
+EMBED_WORKFLOW_PK=your_publishable_key_here
+NEXT_PUBLIC_EMBED_WORKFLOW_PK=your_publishable_key_here
+EMBED_WORKFLOW_UI_VERSION=latest
```
here too

amrani replied:
we should replace "latest" with "x.x.x"

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,667 @@
+
+**User-Specific Workflows:** Each user has their own set of workflows and automations. The app uses `EMBED_WORKFLOW_DEFAULT_USER` (defaults to 'main') to determine which user's workflows to load. This is important because:
+
+- Each user can have different workflows and triggers
```
remove "and triggers"

Triggers are defined at the account level. But yes, triggers can be set for user-groups, though, I'd remove to avoid confusion.

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,667 @@
+
+- Each user can have different workflows and triggers
+- Tasks and automations are isolated per user
+- In a multi-tenant scenario, you'd set different users for different clients
```
this line about multi-tenancy might also lead to confusion. Let's leave this out too

AntonyWilliam committed:
Update with PR fixes

AntonyWilliam committed:
update

AntonyWilliam committed:
IUpdate

amrani commented on beginner/todo-slack-notifications/README.md:
````diff
@@ -0,0 +1,676 @@
+**1. Trigger API Call:**
+Your application sends a POST request to trigger workflows:
+```javascript
+const response = await fetch('/api/trigger-workflow', {
````
did you test this? This is not an endpoint for our API

AntonyWilliam replied:
It was a mistake of mine. I forgot to update it, i'm going to be more careful from now on.

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,676 @@
+  // For testing purposes, you can use a temporary user token instead of a JWT.
+  // Note: This token will expire - you'll need to refresh the page and update the token when it does.
+  // EWF.load("pk_live_XzF0WUAKE2XM6N3O8Y6VeckS", { userToken: "eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaEpJbHBuYVdRNkx5OWhZM1IxYldsNlpTOURiMjV1WldOMFFuazZPbFZ6WlhJdk1ERTVOemRsWVRFdFlUZGxNaTAzTVdGaUxXRTVPV1V0WTJNNVl6TTRZV0l6TlRZelAyVjRjR2x5WlhOZmFXNDlNVGN5T0RBd0Jqb0dSVlE9IiwiZXhwIjoiMjAyNS0wNy0wMVQxNjozMzo0Mi42MDhaIiwicHVyIjoiZGVmYXVsdCJ9fQ==--1d095a56e1ad4346325c781f7790ee63a8c3d099" });
+</script>
```
please do not copy and paste user Tokens

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,676 @@
+<div class="EWF__app" data-base-path="workflows"></div>
+
+<script type="text/javascript">
+  EWF.load("pk_live_XzF0WUAKE2XM6N3O8Y6VeckS", { jwt: "REPLACE_ME_WITH_YOUR_JWT" });
```
use "pk_live_1234" for all publishable keys.

AntonyWilliam replied:
Understood!

amrani commented on beginner/todo-slack-notifications/README.md:
```diff
@@ -0,0 +1,676 @@
+<!-- Load CSS / JS -->
+<!-- Check latest version at: https://embedworkflow.com/ui-version -->
+<link rel="stylesheet" media="screen" href="https://cdn.ewf.to/ewf-latest.css">
+<script src="https://cdn.ewf.to/ewf-latest.js"></script>
```
ewf-latest.js is not a real url. 

We should be using the environment variable here for the current version. Let me know if you need help how

AntonyWilliam replied:
Huge thanks, David! I updated it

amrani commented on beginner/todo-slack-notifications/README.md:
````diff
@@ -0,0 +1,676 @@
+*Sent from Todo App Workflow*
+```
+
+## Customization
````
do we even need this section?

AntonyWilliam replied:
My first idea was to give ideas of personalization to the users, but I believe that we should focus on keeping the content shorter

AntonyWilliam committed:
Update + fixes
